rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,321 @@
+   ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>):
+    %add = stablehlo.add %arg1, %arg2 : tensor<i32>
+    "stablehlo.return"(%add) : (tensor<i32>) -> ()
+  }) {dimensions = dense<[0]> : tensor<1xi64>}
+    : (tensor<?xi32, #stablehlo.type_extensions<bounds = [4]>>,
+       tensor<i32>) -> tensor<i32>
+  func.return %sum : tensor<i32>
+}
+```
+
+## (P4) Comparison with unbounded dynamic ops
+
+In addition to `set_dimension_size` and `get_dimension_size` ops, StableHLO
+producers may also use unbounded dynamic ops like `real_dynamic_slice` and
+`dynamic_pad` to perform operations on dynamically shaped tensors. For example,
+the above `@dynamic_sum` computation can be performed by using the
+`real_dynamic_slice` op instead of the `set_dimension_size` op. With that, the
+above example can be rewritten as,
+
+```mlir
+func.func @dynamic_sum(%data: tensor<4xi32>, %batch_size: tensor<i32>) -> tensor<i32> {
+  %start = stablehlo.constant dense<0> : tensor<1xi32>
+  %limit = stablehlo.reshape %batch_size : (tensor<i32>) -> tensor<1xi32>
+  %strides = stablehlo.constant dense<1> : tensor<1xi32>
+  %dynamic_data =  stablehlo.real_dynamic_slice %data, %start, %limit, %strides
+    : (tensor<4xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>)
+      -> tensor<?xi32>
+
+  %zero = stablehlo.constant dense<0> : tensor<i32>
+  %sum = "stablehlo.reduce"(%dynamic_data, %zero) ({
+   ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>):
+    %add = stablehlo.add %arg1, %arg2 : tensor<i32>
+    "stablehlo.return"(%add) : (tensor<i32>) -> ()
+  }) {dimensions = dense<[0]> : tensor<1xi64>}
+    : (tensor<?xi32>, tensor<i32>) -> tensor<i32>
+  func.return %sum : tensor<i32>
+}
+```
+
+Originally, XLA HLO introduced `set_dimension_size` op as it neither had dynamic
+types nor dynamic ops. StableHLO dialect doesn't have these limitations and
+therefore new users don't need to make use of this low level op. The TensorFlow
+and JAX teams believe that this hypothesis should be correct based on their
+experiences so far. Dynamic operations `real_dynamic_slice` and `dynamic_pad`
+can be used instead.
+
+The following example demonstrates the differences between programs using
+unbounded dynamism and bounded dynamism.
+
+```mlir
+func.func @slice_with_unbounded_dynamism(%data: tensor<7xf32>, %start: tensor<1xi32>, %limit: tensor<1xi32>) -> tensor<?xf32> {
+  %strides = stablehlo.constant dense<1> : tensor<1xi32>
+  %result =  stablehlo.real_dynamic_slice %data, %start, %limit, %strides
+    : (tensor<7xf32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>)
+      -> tensor<?xf32>
+  func.return %result : tensor<?xf32>
+}
+```
+
+```mlir
+func.func @slice_with_bounded_dynamism(%data: tensor<7xf32>, %start: tensor<1xi32>, %limit: tensor<1xi32>) -> tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>> {
+  %zero = stablehlo.constant dense<0> : tensor<1xi32>
+  %size = stablehlo.constant dense<7> : tensor<1xi32>
+
+  // Conditionally add size of the input data if the indices are negative.
```

Comment:
not really. this example was assuming that real_dynamic_slice also allows negative indices. current semantics are not clear to me. Removed it to avoid any confusion.